File: src/tokenizer/destroy_tokens.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   destroy_tokens.c                                   :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jschwabe <jschwabe@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/03/29 20:07:10 by jschwabe          #+#    #+#             */
/*   Updated: 2024/03/30 11:29:11 by jschwabe         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "struct.h"
#include "tokens.h"
#include "libutils.h"

void	destroy_all_tokens(t_shell *shell)
{
	t_token	*token;
	size_t	i;
	size_t	ii;

	token = shell->token;
	i = 0;
	while (token && i < shell->token_len)
	{
		if (token[i].cmd_args)
		{
			ii = 0;
			while (token[i].cmd_args[ii].elem)
			{
				free_null(&token[i].cmd_args[ii].elem);
				ii++;
			}
			free_null(&token[i].cmd_args);
		}
		if (token[i].bin)
			free_null(&(token[i].bin));
		i++;
	}
	free_null(&shell->token);
	shell->token_len = 0;
}

File: src/tokenizer/destroy_tokens.rs
use ::libc;

use crate::{t_shell, t_token};

#[unsafe(no_mangle)]
pub unsafe fn destroy_all_tokens(shell: &mut t_shell) {
	let mut i = 0;
	if shell.token_len.is_some() {
		while i < shell.token_len.unwrap() {
			if !((*shell.token.add(i)).cmd_args_vec).is_empty() {
				// let mut ii = 0;
				// while ii < (*shell.token.add(i)).cmd_args_vec.len() {
				// 	free_null(
				// 		&mut (*shell.token.add(i)).cmd_args_vec[ii].elem as *mut *mut libc::c_char
				// 			as *mut libc::c_void,
				// 	);
				// 	ii += 1;
				// }
			}
			i += 1;
		}
	}
	{
		let p = &mut shell.token as *mut *mut t_token as *mut libc::c_void;
		if p.is_null() {
			return;
		}
		let ptr: *mut *mut libc::c_void = p as *mut *mut libc::c_void;
		if !(*ptr).is_null() {
			libc::free(*ptr);
		}
		*ptr = std::ptr::null_mut::<libc::c_void>();
	};
	shell.token_len = None;
}
